libraries/tlv-account-resolution/README.md

```diff
@@ -39,21 +39,112 @@ ExtraAccountMetas::init_with_account_metas::<MyInstruction>(&mut buffer, &extra_
-let remaining_account_infos = &[]; 
+let remaining_account_infos = &[];
 ExtraAccountMetas::add_to_cpi_instruction::<MyInstruction>(
     &mut cpi_instruction,
     &mut cpi_account_infos,
     &buffer,
     &remaining_account_infos,
+    None,
+).unwrap();
+```
+
+If you want to store information about required additional accounts that have a Program-Derived Address (PDA), thus their address may not be known at the time of packing the account data, you can do so by modifying the above example to look like this:
+
+```rust
+use {
+    solana_program::{account_info::AccountInfo, instruction::{AccountMeta, Instruction}, pubkey::Pubkey},
+    spl_type_length_value::discriminator::{Discriminator, TlvDiscriminator},
+    spl_tlv_account_resolution::state::ExtraAccountMetas,
+};
+
+struct MyInstruction;
+impl TlvDiscriminator for MyInstruction {
+    const TLV_DISCRIMINATOR: Discriminator = Discriminator::new([1; Discriminator::LENGTH]);
+}
+
+// Notice the use of `into()` for the type `AccountMeta`, and we're building this array of `RequiredAccount`
+let required_accounts = [
+    AccountMeta::new(Pubkey::new_unique(), false).into(),
+    AccountMeta::new(Pubkey::new_unique(), true).into(),
+    AccountMeta::new_readonly(Pubkey::new_unique(), true).into(),
+    AccountMeta::new_readonly(Pubkey::new_unique(), false).into(),
+    RequiredAccount::Pda {
+        seeds: vec![
+            Seed::Lit,
+            Seed::Arg(SeedArgType::U8),
+            Seed::Arg(SeedArgType::String),
+        ],
+        is_signer: false,
+        is_writable: true,
+    },
+    RequiredAccount::Pda {
+        seeds: vec![
+            Seed::Lit,
+            Seed::Arg(SeedArgType::Pubkey),
+            Seed::Arg(SeedArgType::Pubkey),
+            Seed::Arg(SeedArgType::Pubkey),
+        ],
+        is_signer: false,
+        is_writable: true,
+    },
+];
+
+let account_size = ExtraAccountMetas::size_of(extra_metas.len()).unwrap();
+let mut buffer = vec![0; account_size];
+
+// Initialize with "required accounts" instead of `AccountMeta` or `AccountInfo`
+ExtraAccountMetas::init_with_required_accounts::<MyInstruction>(
+    &mut buffer,
+    &required_accounts,
+)
+.unwrap();
+
+// On the execution side of things, we have to provide the seeds we used to build the Program-Derived Addresses that correspond to each stored required PDA account in the TLV structure
+let provided_seeds = vec![
+    SeedConfig::new(("SomeSeed", 1u8, String::from("SomeSeed"))),
+    SeedConfig::new((
+        "SomeSeed",
+        Pubkey::new_unique(),
+        Pubkey::new_unique(),
+        Pubkey::new_unique(),
+    )),
+];
+
+// Then you provide the seeds as you add accounts to the instruction (off-chain)
+let program_id = Pubkey::new_unique();
+let mut instruction = Instruction::new_with_bytes(program_id, &[], vec![]);
+ExtraAccountMetas::add_to_instruction::<MyInstruction>(&program_id, &mut instruction, &buffer, Some(provided_seeds)).unwrap();
```

Comment:
We can potentially simplify this tremendously by *enforcing* that PDAs must be derivable by:

* the account pubkeys already in the instruction
* the literal account pubkeys in the ExtraAccountMetas
* the instruction data bytes already there
* any literal data stored in its entry

This way, we never need to pass `provided_seeds`, and the Seeds types become:

```
enum Seeds {
    InstructionPubkey(u8), // index in the instruction
    ExtraAccountPubkey(u8), // index in `ExtraAccountMetas`
    InstructionBytes(u8, u8), // a range to use from the account bytes
    Literal(Vec<u8>), // not actually a vec, but some number of encoded bytes
}
```

Naturally, you'll need to do something more clever for the literal, maybe you can simplify this by making different enum variants for different sizes, ie

```
enum Seeds {
    ... everything else ...
    LiteralByte([u8; 1]),
    LiteralTwoBytes([u8; 2]),
    LiteralThreeBytes([u8; 3]),
   ... etc, probably up to 16 or so?
}
```

What do you think?

Reply:
Yes, I agree! This was the direction I realized this needed to go once I finally worked my way out to the client-side calls and token program(s). 

With the approach you're suggesting, we can implement account resolution exactly the way anchor's SDK does it, where you almost don't even notice it's happening.

This will be my next step - and that will bring this out of draft phase if we still want this supported

Reply:
In fact, your enum closely resembles the `seeds` in both Anchor's and Shank's IDLs

Reply:
Sweet, I think this is definitely worth adding.

Reply:
Copy that, will circle back on this one